FirebasePerformance/Sources/FPRNanoPbUtils.m

```diff
@@ -0,0 +1,452 @@
+  firebase_perf_v1_IosApplicationInfo iosAppInfo = firebase_perf_v1_IosApplicationInfo_init_default;
+  NSBundle *mainBundle = [NSBundle mainBundle];
+  iosAppInfo.bundle_short_version =
+      FPREncodeString([mainBundle infoDictionary][@"CFBundleShortVersionString"]);
+  iosAppInfo.sdk_version = FPREncodeString([NSString stringWithUTF8String:kFPRSDKVersion]);
+  iosAppInfo.network_connection_info.network_type = FPRNetworkConnectionInfoNetworkType();
+#ifdef TARGET_HAS_MOBILE_CONNECTIVITY
+  CTTelephonyNetworkInfo *networkInfo = NetworkInfo();
+  CTCarrier *provider = networkInfo.subscriberCellularProvider;
+  NSString *mccMnc = FPRValidatedMccMnc(provider.mobileCountryCode, provider.mobileNetworkCode);
+  if (mccMnc) {
+    iosAppInfo.mcc_mnc = FPREncodeString(mccMnc);
+  }
+  if (iosAppInfo.network_connection_info.network_type ==
+      firebase_perf_v1_NetworkConnectionInfo_NetworkType_MOBILE) {
+    iosAppInfo.network_connection_info.mobile_subtype = FPRCellularNetworkType();
+  }
+#endif
+  appInfoMessage.ios_app_info = iosAppInfo;
+
+  NSDictionary<NSString *, NSString *> *attributes =
+      [[FIRPerformance sharedInstance].attributes mutableCopy];
+  struct _firebase_perf_v1_ApplicationInfo_CustomAttributesEntry *custom_attributes =
+      calloc(attributes.count, sizeof(firebase_perf_v1_ApplicationInfo_CustomAttributesEntry));
+  appInfoMessage.custom_attributes_count = attributes.count;
+  __block NSUInteger index = 0;
+  [attributes enumerateKeysAndObjectsUsingBlock:^(NSString *key, NSString *value, BOOL *stop) {
+    custom_attributes[index].key = FPREncodeString(key);
+    custom_attributes[index].value = FPREncodeString(value);
+    index++;
+  }];
+  appInfoMessage.custom_attributes = custom_attributes;
+
+  return appInfoMessage;
+}
+
+firebase_perf_v1_TraceMetric GetTraceMetric(FIRTrace *trace) {
+  firebase_perf_v1_TraceMetric traceMetric = firebase_perf_v1_TraceMetric_init_default;
+  traceMetric.name = FPREncodeString(trace.name);
+
+  // Set if the trace is an internally created trace.
+  traceMetric.is_auto = trace.isInternal;
+
+  // Convert the trace duration from seconds to microseconds.
+  traceMetric.duration_us = trace.totalTraceTimeInterval * USEC_PER_SEC;
+
+  // Convert the start time from seconds to microseconds.
+  traceMetric.client_start_time_us = trace.startTimeSinceEpoch * USEC_PER_SEC;
+
+  // TODO(visum) Enable counters
+  //  traceMetric.counters = FPRGetProtoCounterForDictionary(trace.counters);
+
+  // TODO(visum) Enable subtraces
+  //  NSMutableArray<FPRMSGTraceMetric *> *subtraces = [[NSMutableArray alloc] init];
+  //  [trace.stages
+  //      enumerateObjectsUsingBlock:^(FIRTrace *_Nonnull stage, NSUInteger idx, BOOL *_Nonnull
+  //      stop) {
+  //        [subtraces addObject:FPRGetTraceMetric(stage)];
+  //      }];
+  //  traceMetric.subtracesArray = subtraces;
+
+  // TODO(visum) Enable custom attributes
```

Comment:
The next couple of lines can be cleaned up.

Reply:
Updated.